### PR TITLE
Fast exchange

### PR DESCRIPTION
**Pull request by gitpeterwind**
Exchange (saved, i.e. not on the fly) is twice as fast for large systems. Improvement in the parallelization setup:
- Division into "tasks" and dynamical allocation of task between "workers"
- One task involves several orbital, so that they can be reused (saving communication time)

More upgrade of Bank:
- One of the "bank" process is reserved as "task manager", for dynamical allocation of tasks.
- "id" are no more used as MPI tags, allowing to use any int (it had previously been limited by the mpi's max tag value)
- some refactoring

An example of setup for Betzy, where the bank processes do not "occupy" a full worker process, but only one core instead. This allows to better take advantage of all the cores on a node.

### CODE REVIEW

**Comment by gitpeterwind** [comment]
@stigrj , What was missing in the exchangeD1 for full functionality of the not-on-fly setup?

**Comment by stigrj** [comment]
I don't remember. Is it not working?

**Comment by stigrj** [comment]
You said a month ago that the non-diagonal terms were not computed

**Comment by gitpeterwind** [comment]
It is working for normal HF, but in a commit this summer PR you switched it off for some reason. For example what is "dagger"?

**Comment by stigrj** [comment]
Check this discussion: https://github.com/gitpeterwind/mrchem/pull/2

**Comment by stigrj** [comment]
Seems the problem was mixing several OrbitalVectors in the bank at the same time, but this should be seamless now with the new BankAccounts. `dagger()` is the adjoint operator. For response (D2) this means swapping `x_i` and `y_i` terms:
```
K_1.apply(phi_p)  = sum_i   x_i * P[phi_i.dagger() * phi_p] +
                    sum_i phi_i * P[  y_i.dagger() * phi_p]  
K_1.dagger(phi_p) = sum_i phi_i * P[  x_i.dagger() * phi_p] +
                    sum_i   y_i * P[phi_i.dagger() * phi_p]
```

**Comment by gitpeterwind** [comment]
Ah, thanks maybe I start to understand: K.dagger is actually more "K.dagger.apply"?

File: doc/users/betzy_example.job [hunk 1]
```diff
@@ -1,9 +1,8 @@
...
-#SBATCH --tasks-per-node=8
-#SBATCH --cpus-per-task=16
+#SBATCH --tasks-per-node=12
 
 export UCX_LOG_LEVEL=ERROR
-export OMP_NUM_THREADS=16
+export OMP_NUM_THREADS=15
 
-~/my_path/to/mrchem --launcher='mpirun --rank-by node --map-by numa --bind-to numa' mw6
+~/my_path/to/mrchem --launcher='mpirun --rank-by node --map-by numa --bind-to numa --oversubscribe' mw6
```
Comment (stigrj) [inline doc/users/betzy_example.job:8]: The launching procedure is getting complicated. We need to make sure it's transferable to other MPI runtimes and machines like Saga where we don't necessarily get the same number of tasks on each node.

Comment (gitpeterwind) [reply doc/users/betzy_example.job:8]: Yes, this is only for Betzy. 
Note that any "normal" launch will work too, but will not necessarily fully use the resources of the node. ... in this sense this should more be considered as an optimization setting.

**Comment by gitpeterwind** [comment]
The test fails, but the results difference are within world_prec. The new code will compute the exchange terms in a different order, so it is not so surprising. Does this mean I should update the test results?

**Comment by stigrj** [comment]
I can have a look at the tests. Is this otherwise ready?

**Comment by stigrj** [comment]
Will the exchange terms be computed in arbitrary (not reproducible) order?

**Comment by gitpeterwind** [comment]
When run on one MPI, it will be reproducible. (it is very hard to make a code that is not!).
What is there is ready, but I thought having a further look at this dagger.

**Comment by stigrj** [comment]
Is it possible to force it reproducible with `numerically_exact = true`?

**Comment by gitpeterwind** [comment]
You mean that it will be the same, even with many mpi?
(If it is useful I could try, but it means a very different approach: for example forcing one mpi to do all the work, and things like that)

**Comment by stigrj** [comment]
It has proved useful in the past, but it has to make sense also in the code. The changes in the HF energies seems to have been to the better, so I will update the test references.

**Comment by gitpeterwind** [comment]
I was certainly wrong about how complicated it was to include the "numerically exact" flag, one extra line :)

**Comment by stigrj** [comment]
I get this error if I run with the "old" setup `--nodes=1 --ntasks-per-node=8 --cpus-per-task=16` and `mpirun --map-by numa --bind-to numa` on Betzy:
```
Error: /cluster/home/stig/src/mrchem/src/utils/Bank.cpp: openAccount(), line 607: Account id mismatch!
```

**Comment by gitpeterwind** [comment]
Could you give me your .inp file?

**Comment by stigrj** [comment]
`/cluster/projects/nn4654k/stig/benzene.inp`

**Comment by gitpeterwind** [comment]
The id bug was when bank_size=2. In that case the task manager would be run by one of (2) the bank in order not to occupy half of the banks, but that was not tested properly.

**Comment by gitpeterwind** [comment]
For H2O there are no real differences between the precomputed and non-precomputed results. but for Benzene there is a difference is due to different approximation in the integral in the non-symmetric case. I have now forced the two to give the same results.

**Comment by gitpeterwind** [comment]
I think it is ready now

**Review by stigrj — APPROVED** [review]

**Pull request merged**